### PR TITLE
Markov Chain Monte Carlo Gaussian splat optimizer

### PR DESCRIPTION
**Pull request by fwilliams**
Had issues pushing to @harrism's PR branch so making a new one here. This is mostly his work that I took over during the holidays.

Closes https://github.com/openvdb/fvdb-reality-capture/issues/181
Closes https://github.com/openvdb/fvdb-reality-capture/issues/182
Closes https://github.com/openvdb/fvdb-reality-capture/issues/183
Part of https://github.com/openvdb/fvdb-reality-capture/issues/180

This pull request introduces support for a new MCMC-based optimizer for Gaussian Splat radiance field reconstruction, refactors the optimizer registration and deserialization system for extensibility, and adds shared test utilities for Gaussian Splat models. The changes enable users to select between classic and MCMC optimizer strategies from the frgs CLI, ensure optimizer classes are registered and deserialized robustly, and provide reusable testing infrastructure.

It is now possible to train (for example) the garden scene with the MCMC optimizer with a target maximum number of Gaussians of 3M via the following example command. (To train with the "classic" optimizer just change reconstruct-mcmc to reconstruct).

frgs reconstruct-mcmc ./data/360_v2/garden/ --opt.max_gaussians 3000000

### New MCMC Optimizer Support

 - Added GaussianSplatOptimizerMCMC and GaussianSplatOptimizerMCMCConfig classes, and exposed them in the package API (init.py) for use in reconstruction workflows. [[1]](https://github.com/openvdb/fvdb-reality-capture/pull/203/files#diff-2759820a1a42f599112b12750aab136f7c36cbaf9b52cbc188ed359046b0d878R13-R16) [[2]](https://github.com/openvdb/fvdb-reality-capture/pull/203/files#diff-2759820a1a42f599112b12750aab136f7c36cbaf9b52cbc188ed359046b0d878R37-R38) [[3]](https://github.com/openvdb/fvdb-reality-capture/pull/203/files#diff-9de504483e04f918fb31b67eddc32cbc90c5e2d9353081245e4fbafd9b752912R34-R37)
 - Introduced the ReconstructMCMC CLI command (frgs reconstruct-mcmc), allowing users to reconstruct scenes using the MCMC optimizer strategy via the frgs CLI. [[1]](https://github.com/openvdb/fvdb-reality-capture/pull/203/files#diff-699808f45eb9286c8cf11740a9eb4479fdf785a555166a3753d492cf9c00e142L15-R33) [[2]](https://github.com/openvdb/fvdb-reality-capture/pull/203/files#diff-445e9a7b0b8951b65c2c79fa692acadeb29bd16408c53099fc912d5e4b74ac7aR355-R367)
 - Updated the reconstruction pipeline to select the optimizer type (classic or MCMC) based on the optimizer config provided, and improved docstrings to reflect this. [[1]](https://github.com/openvdb/fvdb-reality-capture/pull/203/files#diff-9de504483e04f918fb31b67eddc32cbc90c5e2d9353081245e4fbafd9b752912L404-R411) [[2]](https://github.com/openvdb/fvdb-reality-capture/pull/203/files#diff-9de504483e04f918fb31b67eddc32cbc90c5e2d9353081245e4fbafd9b752912L439-R446)
 - Python tests for the MCMC optimizer and for training using it.
 - Basic MCMC unit benchmarks

### Optimizer Registration and Serialization Refactor

 - Implemented a splat_optimizer decorator and optimizer registry to ensure all optimizer classes are registered for (de)serialization, and refactored from_state_dict to use the registry for robust deserialization. [[1]](https://github.com/openvdb/fvdb-reality-capture/pull/203/files#diff-6834ff0b5a5c8f48f5f900b99ed130ed9923ae6a97d5057ed18988a729ce7f8dL5-R36) [[2]](https://github.com/openvdb/fvdb-reality-capture/pull/203/files#diff-6834ff0b5a5c8f48f5f900b99ed130ed9923ae6a97d5057ed18988a729ce7f8dR49-R57) [[3]](https://github.com/openvdb/fvdb-reality-capture/pull/203/files#diff-6834ff0b5a5c8f48f5f900b99ed130ed9923ae6a97d5057ed18988a729ce7f8dL35-R78) [[4]](https://github.com/openvdb/fvdb-reality-capture/pull/203/files#diff-c0c22eb28bebcf7b745fbbc5d31893e0e4682be84e551e3e16d1051f731056c4L19-R19) [[5]](https://github.com/openvdb/fvdb-reality-capture/pull/203/files#diff-c0c22eb28bebcf7b745fbbc5d31893e0e4682be84e551e3e16d1051f731056c4R251-R259) [[6]](https://github.com/openvdb/fvdb-reality-capture/pull/203/files#diff-c0c22eb28bebcf7b745fbbc5d31893e0e4682be84e551e3e16d1051f731056c4R509)
 - Updated serialization to include the optimizer class name in the state dict, supporting round-trip saving/loading across optimizer types. [[1]](https://github.com/openvdb/fvdb-reality-capture/pull/203/files#diff-c0c22eb28bebcf7b745fbbc5d31893e0e4682be84e551e3e16d1051f731056c4R509) [[2]](https://github.com/openvdb/fvdb-reality-capture/pull/203/files#diff-9de504483e04f918fb31b67eddc32cbc90c5e2d9353081245e4fbafd9b752912L560-R563)

### Testing Infrastructure Improvements

 - Added a tests/init.py to make the test directory a package, enabling shared test helpers.
 - Introduced reusable Gaussian Splat test utilities in tests/unit/common.py, including scene loading, model initialization, and rendering helpers, as well as a mixin for test cases.

### CODE REVIEW

**Comment by swahtz** [comment]
> it's not possible to train (for example) the garden scene with the MCMC optimizer with a target maximum number of Gaussians of 3M via the following example command. (To train with the "classic" optimizer just change reconstruct-mcmc to reconstruct).
> 
> frgs reconstruct-mcmc ./data/360_v2/garden/ --opt.max_gaussians 3000000

I'm confused.  If it's not possible to optimize that scene with the MCMC optimizer, why?  We should change to using the classic optimizer under what circumstances?

File: fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer_mcmc.py [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+              or :func:`from_state_dict`.
+    """
+
+    __PRIVATE__ = object()
+
+    def __init__(
+        self,
+        model: GaussianSplat3d,
+        config: GaussianSplatOptimizerMCMCConfig,
+        optimizer: torch.optim.Adam,
```
Comment (swahtz) [inline fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer_mcmc.py]: **❓ question:** ‏Does this always have to be a `torch.optim.Adam`?  Can it be more generic as a `torch.optim.Optimizer` class?

File: fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer_mcmc.py [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+                break
+        if means_lr is None:
+            raise RuntimeError("Could not find 'means' param group in optimizer")
+        noise_scale = float(self._config.noise_lr) * means_lr
+        if noise_scale != 0.0:
+            self._model.add_noise_to_means(noise_scale=noise_scale)
+
+        self._optimizer.step()
+        self._step_count += 1
+        # Decay the means learning rate
```
Comment (swahtz) [inline fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer_mcmc.py]: **💡 suggestion:** ‏For clarity/modularity, shouldn't this logic exist in a learning rate scheduler (`torch.optim.LRScheduler`)?  That way this logic might fit in better with existing paradigms and let people change this LR decay logic.

**Comment by harrism** [comment]
> > it's not possible to train (for example) the garden scene with the MCMC optimizer with a target maximum number of Gaussians of 3M via the following example command. (To train with the "classic" optimizer just change reconstruct-mcmc to reconstruct).
> > frgs reconstruct-mcmc ./data/360_v2/garden/ --opt.max_gaussians 3000000
> 
> I'm confused. If it's not possible to optimize that scene with the MCMC optimizer, why? We should change to using the classic optimizer under what circumstances?

This was a typo. Now fixed: `s/not/now`.

File: fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer.py [hunk 3]
```diff
@@ -636,6 +646,15 @@ def update_state_function(x: torch.Tensor):
         )
         return {"num_duplicated": num_duplicated, "num_split": num_split, "num_deleted": num_deleted}
 
+    def regularization_loss(self) -> torch.Tensor:
+        """
+        No-op regularization loss for the basica optimizer.
```
[the remark this answers is not in the document]
Comment (harrism) [reply fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer.py]: Fixed.

File: fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer_mcmc.py [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+    """
+
+    insertion_rate: float = 1.05
+    """
+    The rate at which new Gaussians are inserted per step.
+
+    Default: ``1.05`` (i.e., 5% more Gaussians per refinement step).
+    """
+
+    n_max: int = 51
```
[the remark this answers is not in the document]
Comment (harrism) [reply fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer_mcmc.py]: Fixed

File: fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer_mcmc.py [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+            self._binomial_coeffs = self._build_binomial_coeffs(n_max=n_max, device=self._model.device)
+            self._binomial_coeffs_n_max = n_max
+        return self._binomial_coeffs, n_max
+
+    @staticmethod
+    def _build_binomial_coeffs(n_max: int, device: torch.device) -> torch.Tensor:
+        """
+        Build a binomial coefficients lookup table of shape [n_max, n_max] as float32.
+        Matches the reference implementation used in fvdb-core tests.
+        """
```
[the remark this answers is not in the document]
Comment (harrism) [reply fvdb_reality_capture/radiance_fields/gaussian_splat_optimizer_mcmc.py]: Fixed.
